bin/full-node/src/run/network_service.rs

```diff
@@ -790,68 +602,620 @@ pub enum InitError {
+                    let _ = lock
+                        .blocks_requests
+                        .remove(&request_id)
+                        .unwrap()
+                        .send(response);
+                }
+                service::Event::GrandpaWarpSyncRequestResult { .. }
+                | service::Event::StateRequestResult { .. }
+                | service::Event::StorageProofRequestResult { .. }
+                | service::Event::CallProofRequestResult { .. }
+                | service::Event::KademliaFindNodeRequestResult { .. } => {
+                    // We never start a request of this kind.
+                    unreachable!()
+                }
+                service::Event::RequestInCancel { .. } => {
+                    // Requests are answered immediately, and thus cancelling events can't happen.
+                    unreachable!()
+                }
+                service::Event::KademliaDiscoveryResult {
+                    operation_id,
+                    result,
+                } => {
+                    let chain_index = lock
+                        .kademlia_discovery_operations
+                        .remove(&operation_id)
+                        .unwrap();
+                    match result {
+                        Ok(nodes) => {
+                            tracing::debug!(discovered = ?nodes, "discovered");
+                            lock.network.discover(&Instant::now(), chain_index, nodes);
+                        }
+                        Err(error) => {
+                            tracing::debug!(%error, "discovery-error");
+                        }
+                    }
+                }
+                service::Event::IdentifyRequestIn {
+                    peer_id,
+                    request_id,
+                } => {
+                    tracing::debug!(%peer_id, "identify-request");
+                    lock.network.respond_identify(request_id, "smoldot");
+                }
+                service::Event::BlocksRequestIn {
+                    peer_id,
+                    chain_index,
+                    config,
+                    request_id,
+                } => {
+                    tracing::debug!(%peer_id, "incoming-blocks-request");
+                    let mut _jaeger_span = inner.jaeger_service.incoming_block_request_span(
+                        &inner.local_peer_id,
+                        &peer_id,
+                        config.desired_count.get(),
+                        if let (1, protocol::BlocksRequestConfigStart::Hash(block_hash)) =
+                            (config.desired_count.get(), &config.start)
+                        {
+                            Some(block_hash)
+                        } else {
+                            None
+                        },
+                    );
+
+                    // TODO: is it a good idea to await here while the lock is held and freezing the entire networking background task?
```

Comment:
Yeah it's a rhetorical question. The answer is surely no, and this should be fixed at some point but not worth at the moment since the full node is secondary.